### DISCUSSION TITLE
Generate authorization URL server side?

### DISCUSSION THREAD

**Discussion by alessandrojcm** [nextauthjs/next-auth#2719] [Help]
Hi there, we're currently working on migrating our in-house authorization implementation to a third party solution (Fusionauth). For this we've decided to use next-auth for our application.

Now, we want this to be as transparent as possible for our users. Thus, we think that having a home page that just redirects to the OAuth provider would cause confusion. So my question is, is there a way of generating the authorization URL on `getServerSideProps`? (is this even a good idea?). 

I know calling the `/api/auth/signin/fusionauth` endpoint directly from `getServerSideProps` won't work. By reading next-auth code, I can see that this is achieved by using the `getAuthorizationUrl` function, but this is not a public API. So, any ideas on how could we achieve this?

Thanks!

**Comment by balazsorban44** [comment]
Assuming you will have a custom provider, I would rather use

`signIn("Id")`

https://next-auth.js.org/getting-started/client#signin

You could start it from getServerSideProps also though

https://next-auth.js.org/getting-started/rest-api#post-apiauthsigninprovider

this might also be useful
https://github.com/nextauthjs/next-auth/discussions/2621#discussioncomment-1248583

the returned url then can be visited with
the redirect option of getServerSideProps https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering


I wouldn't use getServerSideProps, only if it's absolutely necessary for your usecase though. What do you hope to gain by using it?

**Reply by alessandrojcm** [reply]
Hey, thanks for the insight. We use `getServerSideProps` to check if the user has a valid session and do the appropriate redirections. 

Right now we do as you suggested, using `signIn`. But our problem with that approach is that the users would land on the login page just to press a button to be redirected to the auth provider. 

We think this would be a little bit confusing, since in the current implementation they just land in the login form directly. So my idea was to try to generate the authorization URL and redirect directly to the auth provider. Like I said, I'm not sure if this is even a good idea, hence my question.

**Reply by balazsorban44** [reply]
did you see the link I sent you? 😅 signIn("id") exists exactly for your usecase.

id is the the provider id you define in your provider config

from context, I'll guess you will need `signIn("fusionauth")`

**Reply by alessandrojcm** [reply]
Sorry, I think I didn't explain myself very clearly :sweat_smile: Yes I'm aware of `signIn` (as I said, it is what we're using right now); now, that is a _client side_ function. That means the user will have to: go to the app, land into the "log in" screen, click the sign-in button, be redirected to fusionauth, be redirected back to the application. 

What I want to know is if it's possible to do this redirection on the _server side_ as in: the user goes navigates to the domain, the app checks if there's an active session (using `getServerSideProps` and `getSession`, as we're doing now), and then _redirects directly_ to the auth provider should an active session not be present. Like the authorization URL you get on #2621, but on the server.

**Reply by balazsorban44** [reply]
I see. Yes, I believe I explained that usecase above as well.

Although it feels like a tiny blocking experience from the UX perspective IMO. you could do that session check client side in a useEffect or similar, and just show a skeleton ui while waiting. pretty similar how the vercel.com UI looks like.

(This is personal opinion though)

**Reply by alessandrojcm** [reply]
Alright thanks! To start the flow from getServerSideprops, how should I include the csrf token? As explained in #2621?

**Reply by balazsorban44** [reply]
https://next-auth.js.org/getting-started/client#server-side-example-1

be aware of the import (https://next-auth.js.org/getting-started/upgrade-v4)

I believe you could naively also find it in `ctx.req.cookies` somewhere.

**Reply by alessandrojcm** [reply]
Great thanks for your help, will take a look at that!

**Reply by balazsorban44** [reply]
keep me posted
